src/transport/raw/MessageHeader.h

```diff
@@ -42,38 +42,49 @@
 
 namespace chip {
 
+static constexpr size_t kMaxTagLen = 16;
+
 namespace detail {
 // Figure out the max size of a packet we can allocate, including all headers.
-static constexpr size_t kMaxIPPacketSizeBytes       = 1280;
+static constexpr size_t kMaxUdpIPPacketSizeBytes       = 1280;
+static constexpr size_t kMaxTcpIPPacketSizeBytes       = 12800;
 static constexpr size_t kMaxUDPAndIPHeaderSizeBytes = 48;
 
-static_assert(kMaxIPPacketSizeBytes >= kMaxUDPAndIPHeaderSizeBytes + CHIP_SYSTEM_HEADER_RESERVE_SIZE,
+static_assert(kMaxUdpIPPacketSizeBytes >= kMaxUDPAndIPHeaderSizeBytes + CHIP_SYSTEM_HEADER_RESERVE_SIZE,
               "Matter headers and IP headers must fit in an MTU.");
 
 // Max space we have for our Application Payload and MIC, per spec.
 static constexpr size_t kMaxPerSpecApplicationPayloadAndMICSizeBytes =
-    kMaxIPPacketSizeBytes - kMaxUDPAndIPHeaderSizeBytes - CHIP_SYSTEM_HEADER_RESERVE_SIZE;
+    kMaxUdpIPPacketSizeBytes - kMaxUDPAndIPHeaderSizeBytes - CHIP_SYSTEM_HEADER_RESERVE_SIZE;
 
 // Max space we have for our Application Payload and MIC in our actual packet
 // buffers.  This is the size _excluding_ the header reserve.
 static constexpr size_t kMaxPacketBufferApplicationPayloadAndMICSizeBytes = System::PacketBuffer::kMaxSize;
 
-static constexpr size_t kMaxApplicationPayloadAndMICSizeBytes =
+static constexpr size_t kMaxApplicationUdpPayloadAndMICSizeBytes =
     min(kMaxPerSpecApplicationPayloadAndMICSizeBytes, kMaxPacketBufferApplicationPayloadAndMICSizeBytes);
 
-} // namespace detail
+static constexpr size_t kMaxApplicationTcpPayloadAndMICSizeBytes =
+    kMaxTcpIPPacketSizeBytes - CHIP_SYSTEM_HEADER_RESERVE_SIZE;
 
-static constexpr size_t kMaxTagLen = 16;
-
-static_assert(detail::kMaxApplicationPayloadAndMICSizeBytes > kMaxTagLen, "Need to be able to fit our tag in a message");
+static_assert(detail::kMaxApplicationUdpPayloadAndMICSizeBytes > kMaxTagLen, "Need to be able to fit our tag in a message");
+static_assert(detail::kMaxApplicationTcpPayloadAndMICSizeBytes > kMaxTagLen, "Need to be able to fit our tag in a message");
 
 // This is somewhat of an under-estimate, because in practice any time we have a
 // tag we will not have source/destination node IDs, but above we are including
 // those in the header sizes.
-static constexpr size_t kMaxAppMessageLen = detail::kMaxApplicationPayloadAndMICSizeBytes - kMaxTagLen;
+static constexpr size_t kMaxUdpAppMessageLen = kMaxApplicationUdpPayloadAndMICSizeBytes - kMaxTagLen;
+static constexpr size_t kMaxTcpAppMessageLen = kMaxApplicationTcpPayloadAndMICSizeBytes - kMaxTagLen;
+} // namespace detail
 
 static constexpr uint16_t kMsgUnicastSessionIdUnsecured = 0x0000;
 
+#if CHIP_CONFIG_TCP_SUPPORT_SERVER || CHIP_CONFIG_TCP_SUPPORT_CLIENT
+static constexpr size_t kMaxAppMessageLen = detail::kMaxTcpAppMessageLen;
```

Comment:
There cannot be one max for both TCP and MRP, otherwise it's impossible to use MRP + TCP in the same stack ever. That is, a given controller would be unable to work with MRP based clients and TCP based clients but enforce the right sizes.

Reply:
so do you suggest enforcing that limit from the session perspective ? 
i.e. Defining two maximums (one for MRP and one for TCP) and in the encrypt method for session messages,that limit is enforced based on the transported used in the session ?

Reply:
Well, there is a fundamental issue in that the IM layer has a transport knowledge due to the limits of MRP. Before we start preparing a message with CommandSender, ReadClient, etc, there must be a clean way to layer-cross the knowledge of the underlying session type (e.g. CASE over TCP versus MRP), in such a way that the limit can be selected properly without having to have one global limit that needs to apply to both (which will break MRP if TCP is enabled).

Reply:
FYI @bzbarsky-apple

Reply:
How about having a way to ask a session what the max message size it supports is?

Reply:
Yes, ideally, the max defined for TCP should be the max for TCP only. The UDP max length is much smaller and probably the enforcing checks need to be a tight upper bound. 
Having that information retrievable from the Session seems reasonable. However, not sure how the verification functions in Encrypt() would fetch that. Something via the CryptoContext maybe?